src/pagination/index.tsx

```diff
@@ -0,0 +1,178 @@
+import { create, tsx } from '@dojo/framework/core/vdom';
+import { createICacheMiddleware } from '@dojo/framework/core/middleware/icache';
+import i18n from '@dojo/framework/core/middleware/i18n';
+import theme from '@dojo/framework/core/middleware/theme';
+import { WNode, RenderResult } from '@dojo/framework/core/interfaces';
+import { createResource, DataTemplate } from '@dojo/framework/core/resource';
+
+import Icon from '../icon';
+import Select, { defaultTransform } from '../select';
+
+import bundle from './Pagination.nls';
+import * as css from '../theme/default/pagination.m.css';
+
+function createMemoryTemplate<S = void>(): DataTemplate<S> {
+	return {
+		read: (_, put, get) => {
+			let data: any[] = get();
+			put(0, data);
+			return { data, total: data.length };
+		}
+	};
+}
+
+const memoryTemplate = createMemoryTemplate();
+
+export interface PaginationProperties {
+	/** The initial page number */
+	initialPage?: number;
+
+	/** The initial page size */
+	initialPageSize?: number;
+
+	/** Callback fired when the page number is changed */
+	onPageChange(page: number): void;
+
+	/** Callback fired when the page size is changed */
+	onPageSizeChange?(pageSize: number): void;
+
+	/** Page size options to display in the page size selector */
+	pageSizes?: number[];
+
+	/** Custom renderer for page size selector item */
+	pageSizeRenderer?(pageSize: number): RenderResult;
+
+	/** Number of always visible page links before and after the current page */
+	siblingCount?: number;
+
+	/** Total number of pages */
+	total: number;
+}
+
+interface PaginationCache {
+	initialPage?: number;
+	initialPageSize?: number;
+	currentPage: number;
+	pageSize: number;
+}
+
+const icache = createICacheMiddleware<PaginationCache>();
+const factory = create({ theme, icache, i18n }).properties<PaginationProperties>();
+
+export default factory(function Pagination({ middleware: { theme, icache, i18n }, properties }) {
+	const {
+		initialPage,
+		initialPageSize,
+		onPageChange,
+		onPageSizeChange,
+		pageSizes,
+		pageSizeRenderer,
+		siblingCount = 3,
+		total
+	} = properties();
+	const classes = theme.classes(css);
+	const { messages } = i18n.localize(bundle);
+
+	if (initialPage !== undefined && initialPage !== icache.get('initialPage')) {
+		icache.set('initialPage', initialPage);
+		icache.set('currentPage', initialPage);
+	}
+
+	if (initialPageSize !== undefined && initialPageSize !== icache.get('initialPageSize')) {
+		icache.set('initialPageSize', initialPageSize);
+		icache.set('pageSize', initialPageSize);
+	}
+
+	const page = icache.getOrSet('currentPage', 1);
+	const pageSize = icache.getOrSet('pageSize', 10);
+
+	const pageLink = function(
+		page: number,
+		content?: WNode | WNode[],
+		extraClasses: string = classes.numberedLink,
+		key?: string
+	): WNode {
+		return (
+			<button
```

Comment:
How about moving this inline rather than having it in it's own function. I think that would make it clearer.